src/metabase/models/card.clj

```diff
@@ -140,13 +142,39 @@
       :mbql/query (-> query lib/normalize lib.util/source-card-id)
       nil)))
 
+(defn- card->integer-table-ids
+  "Return integer source table ids for card's :dataset_query."
+  [card]
+  (when-some [query (-> card :dataset_query :query)]
+    (not-empty (filter pos-int? (lib.util/collect-source-tables query)))))
+
+(defn- prefetch-tables-for-cards!
+  "Collect tables from `dataset-cards` and prefetch metadata. Should be used only with metdata provider caching
+  enabled, as per https://github.com/metabase/metabase/pull/45050. Returns `nil`."
+  [dataset-cards]
+  (let [db-id->table-ids (-> (group-by :database_id dataset-cards)
+                             (update-vals (partial into #{} (comp (mapcat card->integer-table-ids)
+                                                                  (remove nil?)))))
+        db-id->mp (into {}
+                        (map (juxt identity lib.metadata.jvm/application-database-metadata-provider))
+                        (keys db-id->table-ids))]
+    (doseq [[db-id table-ids] db-id->table-ids
+            :when (seq table-ids)]
+      (lib.metadata.protocols/metadatas (db-id->mp db-id) :metadata/table table-ids))))
```

Comment:
Defining `db-id->mp` as a separate map is redundant here - each one is only used in one iteration of the `doseq`, which is already grouped by `db-id`.

```suggestion
  (let [db-id->table-ids (-> (group-by :database_id dataset-cards)
                             (update-vals (partial into #{} (comp (mapcat card->integer-table-ids)
                                                                  (remove nil?)))))]
    (doseq [[db-id table-ids] db-id->table-ids
            :let  [mp (lib.metadata.jvm/application-database-metadata-provider db-id)]
            :when (seq table-ids)]
      (lib.metadata.protocols/metadatas mp :metadata/table table-ids))))
```

Reply:
Thanks for the suggestion. Applied!